src/Lib.P2P/SwarmService.cs

```diff
@@ -0,0 +1,1132 @@
+                        connection = await TaskHelper.WhenAnyResultAsync(attempts, cts.Token).ConfigureAwait(false);
+                        cts.Cancel(); // stop other dialing tasks.
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                var attemped = string.Join(", ", possibleAddresses.Select(a => a.ToString()));
+                _log.Trace($"Cannot dial {attemped}");
+                throw new Exception($"Cannot dial {remote}.", e);
+            }
+
+            // Do the connection handshake.
+            try
+            {
+                MountProtocols(connection);
+                IEncryptionProtocol[] security;
+                lock (_protocols)
+                {
+                    security = _protocols.OfType<IEncryptionProtocol>().ToArray();
+                }
+
+                await connection.InitiateAsync(security, cancel).ConfigureAwait(false);
+                await connection.MuxerEstablished.Task.ConfigureAwait(false);
+                Identify1 identify;
+                lock (_protocols)
+                {
+                    identify = _protocols.OfType<Identify1>().First();
+                }
+
+                await identify.GetRemotePeerAsync(connection, cancel).ConfigureAwait(false);
+            }
+            catch (Exception)
+            {
+                connection.Dispose();
+                throw;
+            }
+
+            var actual = Manager.Add(connection);
+            if (actual == connection)
+            {
+                ConnectionEstablished?.Invoke(this, connection);
+            }
+
+            return actual;
+        }
+
+        private async Task<PeerConnection> DialAsync(Peer remote, MultiAddress addr, CancellationToken cancel)
+        {
+            // TODO: HACK: Currenty only the ipfs/p2p is supported.
+            // short circuit to make life faster.
+            if (addr.Protocols.Count != 3
+             || !(addr.Protocols[2].Name == "ipfs" || addr.Protocols[2].Name == "p2p"))
+            {
+                throw new Exception($"Cannnot dial; unknown protocol in '{addr}'.");
+            }
+
+            // Establish the transport stream.
+            Stream stream = null;
+            foreach (var protocol in addr.Protocols)
+            {
+                cancel.ThrowIfCancellationRequested();
+                if (!TransportRegistry.Transports.TryGetValue(protocol.Name, out var transport))
+                {
+                    continue;
+                }
+                
+                stream = await transport().ConnectAsync(addr, cancel).ConfigureAwait(false);
+                if (cancel.IsCancellationRequested)
+                {
+                    stream?.DisposeAsync();
+                    continue;
+                }
+
+                break;
+            }
+
+            if (stream == null)
+            {
+                throw new Exception("Missing a known transport protocol name.");
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)